staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/maplist.go

```diff
@@ -0,0 +1,191 @@
+// keys.
+func (ks *hashKeyStrategy) CompositeKeyFor(obj map[string]interface{}) (interface{}, bool) {
+	const keyDelimiter = "\x00" // 0 byte should never appear in the hash input except as delimiter
+
+	ks.hasher.Reset()
+	for _, key := range ks.sts.XListMapKeys {
+		v, ok := obj[key]
+		if !ok {
+			v = ks.sts.Properties[key].Default.Object
+		}
+
+		switch v.(type) {
+		case bool:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%t", v)
+		case float64:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%f", v)
+		case int64:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%d", v)
+		case string:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%q", v)
+		default:
+			return nil, false // values must be scalars
+		}
+	}
+	return ks.hasher.Sum64(), true
+}
+
+// emptyMapList is a mapList containing no elements.
+type emptyMapList struct{}
+
+func (emptyMapList) get(interface{}) interface{} {
+	return nil
+}
+
+type mapListImpl struct {
+	sts      *schema.Structural
+	ks       keyStrategy
+	elements map[interface{}][]interface{} // composite key -> bucket
+}
+
+func (a *mapListImpl) get(obj interface{}) interface{} {
+	mobj, ok := obj.(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	key, ok := a.ks.CompositeKeyFor(mobj)
+	if !ok {
+		return nil
+	}
+
+	// Scan bucket to handle key collisions and duplicate key sets:
+	var match interface{}
+	for _, element := range a.elements[key] {
+		all := true
+		for _, key := range a.sts.XListMapKeys {
+			va, ok := element.(map[string]interface{})[key]
+			if !ok {
+				va = a.sts.Properties[key].Default.Object
+			}
+
+			vb, ok := mobj[key]
+			if !ok {
+				vb = a.sts.Properties[key].Default.Object
+			}
+
+			all = all && (va == vb)
+		}
+
+		if !all {
+			continue
+		}
+
+		if match != nil {
+			// Duplicate key set / more than one element matches. This condition should
+			// have generated a validation error elsewhere.
+			return nil
+		}
+		match = element
+	}
+	return match // can be nil
+}
+
+func makeKeyStrategy(sts *schema.Structural) keyStrategy {
+	if len(sts.XListMapKeys) == 1 {
+		key := sts.XListMapKeys[0]
+		return &singleKeyStrategy{
+			key:     key,
+			defawlt: sts.Properties[key].Default.Object,
+		}
+	}
+
+	return &hashKeyStrategy{
+		sts: sts,
+	}
+}
+
+// makeMapList returns a queryable interface over the provided x-kubernetes-list-type=map
+// elements. If the provided schema is _not_ an array with x-kubernetes-list-type=map, returns an
+// empty mapList.
+func makeMapList(sts *schema.Structural, ks keyStrategy, items []interface{}) (rv mapList) {
+	if sts.Type != "array" || sts.XListType == nil || *sts.XListType != "map" || len(sts.XListMapKeys) == 0 || len(items) == 0 {
+		return emptyMapList{}
+	}
+
+	elements := make(map[interface{}][]interface{}, len(items))
+
+	for _, item := range items {
+		mitem, ok := item.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		if key, ok := ks.CompositeKeyFor(mitem); ok {
```

Comment:
This is computing all composite keys for all old items at `makeMapList` call-time, even if we'll never use them (if the new object has no items, for example).

Making the existing item key computation lazy in `get` could avoid this.

Depending on what we do for duplicate item scenarios in oldData (I'm sort of inclined to return the first match in a non-unique case), the lazy computation could do one of:

1. fully populate `elements` on first call:

```go
func (a *mapListImpl) get(obj interface{}) interface{} {
	...
	key, ok := a.ks.CompositeKeyFor(mobj)
	if !ok {
		return nil
	}
  if a.elements == nil {
    ... populate a.elements from a.items
  }
	...
}
```

2. populate elements until a matching item is found and short-circuit:

```go
func (a *mapListImpl) get(obj interface{}) interface{} {
	...
	key, ok := a.ks.CompositeKeyFor(mobj)
	if !ok {
		return nil
	}

	// check for already-keyed match
	match, ok := a.elements[key]
	if ok {
		return match
	}

	// key any unkeyed items
	for len(a.unkeyedItems) > 0 {
		// dequeue an unkeyed item
		item := a.unkeyedItems[0]
		a.unkeyedItems = a.unkeyedItems[1:]

		// key the item
		mitem, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		itemKey, ok := a.ks.CompositeKeyFor(mitem)
		if !ok {
			continue
		}
		a.elements[itemKey] = mitem

		// if it matches, short-circuit
		if itemKey == key {
			return mitem
		}
	}

	// no match
	return nil
}
```

Reply:
lazily keying the items SGTM, I've copied over 2nd approach and integrated it.